File: engine/extension-protocol/src/ViewEvent.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum ViewEvent {
    /// Host-generated refresh after an extension invalidates an open view.
    Invalidated,
    Resumed,
    SearchChanged {
        text: String,
    },
    SelectionChanged {
        item_id: Option<String>,
    },
    FilterChanged {
        filter_id: String,
        value: String,
    },
    LoadMore {
        cursor: String,
    },
    ActionInvoked {
        item_id: Option<String>,
        action_id: String,
    },
}

File: apps/desktop/frontend/src/types/ViewEvent.ts
import type { ActionInvocation } from "./ActionInvocation";

export type ViewEvent =
    | { action_id: string; invocation: ActionInvocation; item_id: string | null; kind: "actionInvoked"; }
    | { cursor: string; kind: "loadMore"; }
    | { filter_id: string; kind: "filterChanged"; value: string; }
    | { item_id: string | null; kind: "selectionChanged"; }
    | { kind: "resumed"; }
    | { kind: "searchChanged"; text: string; };
